feat(reminders): threshold filters

ericli3690 opened the pull request:
## Purpose / Description
Add a group of new advanced review reminder options: count new cards, count cards in learning, and count cards in review. When the review reminder is about to send a notification and checks to see if the amount of cards in the deck is greater than the card trigger threshold, it examines these options to check if it should count and consider new cards, cards in learning, and cards in review.

Adds three new checkboxes to the AddEditReminderDialog to toggle these booleans on or off. Edits some logic in NotificationService to add up cards only from selected card type when determining whether the card trigger threshold is met.

Adds three new boolean fields to store the states of these settings to ReviewReminder. Adds unit tests.

## UI
<img width="1080" height="2340" alt="image" src="https://github.com/user-attachments/assets/0ec398b7-4281-4145-b479-13b0891aaa98" />

## Fixes
* GSoC 2025: Review Reminders

## Approach
One of the two planned review reminder advanced options! Just needs some extra checks when handling the `Count` objects.

## How Has This Been Tested?
- Unit tests pass.
- Builds and runs on a physical Samsung S23, API 34.
- Based on my own personal testing, it seems the feature works! I created a few review reminders, reviewed a few cards, etc.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

<!--- Uncomment this section ONLY if this PR introduces new resources (external libraries, icons etc)
## Licenses
_For each new external resource, add a row to the table below:_

| Library | Description | License |
| --- | --- | --- |
| Sample Icon Library | Sample Description | [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt) |

**Maintainers:**

* [ ] Add the https://github.com/ankidroid/Anki-Android/labels/Licenses label
* [ ] Update the [licenses](https://github.com/ankidroid/Anki-Android/wiki/Licences) wiki when merging
--->

ericli3690 commented:
- Rebased.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/AddEditReminderDialogViewModel.kt:
```diff
@@ -98,6 +98,33 @@ class AddEditReminderDialogViewModel(
+                is AddEditReminderDialog.DialogMode.Edit -> dialogMode.reminderToBeEdited.countRev
+            },
+        )
+    val countRev: LiveData<Boolean> = _countRev
```
informational: we're going to be moving to Kotlin 2.3 in February, and this has nicer syntax. You may want to come back to this once it's out

https://kotlinlang.org/docs/whatsnew23.html#explicit-backing-fields

* This PR is blocked on a GitHub CodeQL update: https://github.com/ankidroid/Anki-Android/pull/19945

ericli3690 replied:
Sweet! I've subscribed to the PR and will update this file once it's out. Thanks!

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt:
```diff
@@ -106,9 +107,7 @@ class NotificationServiceTest : RobolectricTest() {
-            addNotes(2).forEach {
-                it.firstCard().update { did = did1 }
-            }
+            addNoteToDeck(did1, count = 2)
```
great job!

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt:
```diff
@@ -150,12 +147,8 @@ class NotificationServiceTest : RobolectricTest() {
-            addNotes(2).forEach {
-                it.firstCard().update { did = did2 }
-            }
+            addNoteToDeck(did1, count = 2)
```
Totally optional:

Feel free to have fun making this even more terse: `addDeck("Deck1").withNotes(2)` etc...

david-allison reviewed (commented):
Code quality is solid!

My main concern is the UI is very wordy.

Have a play around with it, but I'd consider using a `Spannable` with underline and text color, such as this:

<img width="161" height="33" alt="Image" src="https://github.com/user-attachments/assets/7192ad47-1a28-4ac9-abc5-7ef9cfc14b6b" />

(sorry for the screenshot: GitHub didn't support the HTML)

----

optional patch for the tests

```patch
Index: AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt b/AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt
--- a/AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt	(revision d022fa421e976c049e5421143483f1e9af2fc87f)
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt	(date 1768638350193)
@@ -33,6 +33,7 @@
 import com.ichi2.anki.reviewreminders.ReviewReminderCardTriggerThreshold
 import com.ichi2.anki.reviewreminders.ReviewReminderId
 import com.ichi2.anki.reviewreminders.ReviewReminderScope
+import com.ichi2.anki.reviewreminders.ReviewReminderScope.DeckSpecific
 import com.ichi2.anki.reviewreminders.ReviewReminderTime
 import com.ichi2.anki.reviewreminders.ReviewRemindersDatabase
 import com.ichi2.anki.settings.Prefs
@@ -336,10 +337,11 @@
     @Test
     fun `onReceive with rev cards not counted and only rev cards present should not fire notification`() =
         runTest {
-            val did1 = addDeck("Deck", setAsSelected = true)
-            addNoteToDeck(did1, count = 2, queueType = QueueType.Rev)
+            val did1 = addDeck("Deck", setAsSelected = true).withNotes(2, QueueType.Rev)
+
             val reviewReminder = createTestReminder(deckId = did1, countRev = false)
-            ReviewRemindersDatabase.editRemindersForDeck(did1) { mapOf(ReviewReminderId(0) to reviewReminder) }
+
+            ReviewRemindersDatabase.setupDeckReminders(reviewReminder)
 
             triggerDummyReminderNotification(reviewReminder)
 
@@ -349,10 +351,10 @@
     @Test
     fun `onReceive with new cards not counted and only new cards present should not fire notification`() =
         runTest {
-            val did1 = addDeck("Deck")
-            addNoteToDeck(did1, count = 2)
+            val did1 = addDeck("Deck").withNotes(2)
+
             val reviewReminder = createTestReminder(deckId = did1, countNew = false)
-            ReviewRemindersDatabase.editRemindersForDeck(did1) { mapOf(ReviewReminderId(0) to reviewReminder) }
+            ReviewRemindersDatabase.setupDeckReminders(reviewReminder)
 
             triggerDummyReminderNotification(reviewReminder)
 
@@ -362,10 +364,10 @@
     @Test
     fun `onReceive with lrn cards not counted and only lrn cards present should not fire notification`() =
         runTest {
-            val did1 = addDeck("Deck")
-            addNoteToDeck(did1, count = 2, queueType = QueueType.Lrn)
+            val did1 = addDeck("Deck").withNotes(2, QueueType.Lrn)
+
             val reviewReminder = createTestReminder(deckId = did1, countLrn = false)
-            ReviewRemindersDatabase.editRemindersForDeck(did1) { mapOf(ReviewReminderId(0) to reviewReminder) }
+            ReviewRemindersDatabase.setupDeckReminders(reviewReminder)
 
             triggerDummyReminderNotification(reviewReminder)
 
@@ -376,11 +378,12 @@
     fun `onReceive with all cards not counted and many cards present should not fire notification`() =
         runTest {
             val did1 = addDeck("Deck")
-            addNoteToDeck(did1, count = 1, queueType = QueueType.New)
-            addNoteToDeck(did1, count = 1, queueType = QueueType.Lrn)
-            addNoteToDeck(did1, count = 1, queueType = QueueType.Rev)
+                .withNotes(1, QueueType.New)
+                .withNotes(1, QueueType.Lrn)
+                .withNotes(1, QueueType.Rev)
+
             val reviewReminder = createTestReminder(deckId = did1, countNew = false, countLrn = false, countRev = false)
-            ReviewRemindersDatabase.editRemindersForDeck(did1) { mapOf(ReviewReminderId(0) to reviewReminder) }
+            ReviewRemindersDatabase.setupDeckReminders(reviewReminder)
 
             triggerDummyReminderNotification(reviewReminder)
 
@@ -391,10 +394,11 @@
     fun `onReceive with new cards not counted but other kinds present should fire notification`() =
         runTest {
             val did1 = addDeck("Deck")
-            addNoteToDeck(did1, 1, QueueType.New)
-            addNoteToDeck(did1, 1, QueueType.Lrn)
+                .withNotes(1, QueueType.New)
+                .withNotes(1, QueueType.Lrn)
+
             val reviewReminder = createTestReminder(deckId = did1, countNew = false)
-            ReviewRemindersDatabase.editRemindersForDeck(did1) { mapOf(ReviewReminderId(0) to reviewReminder) }
+            ReviewRemindersDatabase.setupDeckReminders(reviewReminder)
 
             triggerDummyReminderNotification(reviewReminder)
 
@@ -407,10 +411,11 @@
     fun `onReceive with new cards not counted and not enough non new cards to trigger threshold should not fire notification`() =
         runTest {
             val did1 = addDeck("Deck")
-            addNoteToDeck(deckId = did1, queueType = QueueType.New)
-            addNoteToDeck(deckId = did1, queueType = QueueType.Lrn)
+                .withNotes(1, QueueType.New)
+                .withNotes(1, QueueType.Lrn)
+
             val reviewReminder = createTestReminder(deckId = did1, thresholdInt = 2, countNew = false)
-            ReviewRemindersDatabase.editRemindersForDeck(did1) { mapOf(ReviewReminderId(0) to reviewReminder) }
+            ReviewRemindersDatabase.setupDeckReminders(reviewReminder)
 
             triggerDummyReminderNotification(reviewReminder)
 
@@ -420,14 +425,12 @@
     @Test
     fun `snooze actions of different notifications and different intervals should be different`() =
         runTest {
-            val did1 = addDeck("Deck1")
-            val did2 = addDeck("Deck2")
-            addNoteToDeck(did1, count = 2)
-            addNoteToDeck(did2, count = 2)
+            val did1 = addDeck("Deck1").withNotes(2)
+            val did2 = addDeck("Deck2").withNotes(2)
+
             val reviewReminderOne = createTestReminder(deckId = did1, thresholdInt = 1)
             val reviewReminderTwo = createTestReminder(deckId = did2, thresholdInt = 1)
-            ReviewRemindersDatabase.editRemindersForDeck(did1) { mapOf(ReviewReminderId(0) to reviewReminderOne) }
-            ReviewRemindersDatabase.editRemindersForDeck(did2) { mapOf(ReviewReminderId(1) to reviewReminderTwo) }
+            ReviewRemindersDatabase.setupDeckReminders(reviewReminderOne, reviewReminderTwo)
 
             val slotOne = slot<Notification>()
             val slotTwo = slot<Notification>()
@@ -490,10 +493,33 @@
     ) = ReviewReminder.createReviewReminder(
         time = ReviewReminderTime(hour = 12, minute = 0),
         cardTriggerThreshold = ReviewReminderCardTriggerThreshold(thresholdInt),
-        scope = if (deckId != null) ReviewReminderScope.DeckSpecific(deckId) else ReviewReminderScope.Global,
+        scope = if (deckId != null) DeckSpecific(deckId) else ReviewReminderScope.Global,
         onlyNotifyIfNoReviews = onlyNotifyIfNoReviews,
         countNew = countNew,
         countLrn = countLrn,
         countRev = countRev,
     )
 }
+
+context(test: NotificationServiceTest)
+private fun DeckId.withNotes(count: Int, queueType: QueueType = QueueType.New): DeckId = this.apply {
+    test.addNoteToDeck(this, count, queueType)
+}
+
+@Suppress("UnusedReceiverParameter")
+private fun ReviewRemindersDatabase.setupDeckReminders(vararg reminders: ReviewReminder) {
+    require(reminders.isNotEmpty()) { "empty list provided" }
+
+    fun ReviewReminder.requireDeckId(): DeckId {
+        require(this.scope is DeckSpecific) { "non-deck-specific reminder provided" }
+        return this.scope.did
+    }
+
+    reminders.withIndex().groupBy { it.value.requireDeckId() }.forEach { (deckId, reminders) ->
+        ReviewRemindersDatabase.editRemindersForDeck(deckId) {
+            reminders.associate {
+                (index, reminder) -> ReviewReminderId(index) to reminder
+            }
+        }
+    }
+}
\ No newline at end of file
```

ericli3690 committed:
feat(reminders): threshold filters

GSoC 2025: Review Reminders

Add a group of new advanced review reminder options: count new cards, count cards in learning, and count cards in review. When the review reminder is about to send a notification and checks to see if the amount of cards in the deck is greater than the card trigger threshold, it examines these options to check if it should count and consider new cards, cards in learning, and cards in review.

Adds three new checkboxes to the AddEditReminderDialog to toggle these booleans on or off, with colored text for the corresponding review state boolean. Edits some logic in NotificationService to add up cards only from selected card type when determining whether the card trigger threshold is met.

Adds three new boolean fields to store the states of these settings to ReviewReminder. Adds unit tests. Modifies some unit test utilities for convenience.

david-allison commented on libanki/testutils/src/main/java/com/ichi2/anki/libanki/testutils/AnkiTest.kt:
```diff
@@ -209,6 +209,33 @@ interface AnkiTest {
+    /**
+     * Convenience method for chaining [addDeck] and [addNoteToDeck].
+     *
+     * Usage: `val deckId = addDeck("My Deck").withNote(count = 5, queueType = QueueType.New)`
```
give naming a second consideration, maybe even define `withNote` to mean 1, and `withNotes` to mean more than 1.

```patch
Subject: [PATCH] 
---
Index: AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt b/AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt
--- a/AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt	(revision f4af44f18bcf1ad2a3348d3ec51a9e8d3b314072)
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/services/NotificationServiceTest.kt	(date 1768720363077)
@@ -87,7 +87,7 @@
     @Test
     fun `onReceive with less cards than card threshold should not fire notification but schedule next`() =
         runTest {
-            val did1 = addDeck("Deck", setAsSelected = true).withNote(count = 2)
+            val did1 = addDeck("Deck", setAsSelected = true).withNotes(count = 2)
             val reviewReminderDeckSpecific = createTestReminder(deckId = did1, thresholdInt = 3)
             val reviewReminderAppWide = createTestReminder(thresholdInt = 3)
             ReviewRemindersDatabase.storeReminders(reviewReminderDeckSpecific, reviewReminderAppWide)
@@ -103,7 +103,7 @@
     @Test
     fun `onReceive with happy path for single deck should fire notification and schedule next`() =
         runTest {
-            val did1 = addDeck("Deck", setAsSelected = true).withNote(count = 2)
+            val did1 = addDeck("Deck", setAsSelected = true).withNotes(count = 2)
             val reviewReminder = createAndSaveDummyDeckSpecificReminder(did1)
 
             triggerDummyReminderNotification(reviewReminder)
@@ -115,8 +115,8 @@
     @Test
     fun `onReceive with happy path for global reminder should fire notification and schedule next`() =
         runTest {
-            addDeck("Deck1").withNote(count = 2)
-            addDeck("Deck2").withNote(count = 2)
+            addDeck("Deck1").withNotes(count = 2)
+            addDeck("Deck2").withNotes(count = 2)
             val reviewReminder = createTestReminder(thresholdInt = 4)
 
             triggerDummyReminderNotification(reviewReminder)
@@ -210,7 +210,7 @@
     @Test
     fun `onReceive with blocked collection should not fire notification but schedule next`() =
         runTest {
-            val did1 = addDeck("Deck", setAsSelected = true).withNote(count = 2)
+            val did1 = addDeck("Deck", setAsSelected = true).withNotes(count = 2)
             val reviewReminderDeckSpecific = createAndSaveDummyDeckSpecificReminder(did1)
             val reviewReminderAppWide = createAndSaveDummyAppWideReminder()
 
@@ -226,7 +226,7 @@
     @Test
     fun `onReceive with snoozed notification should fire notification but not schedule next`() =
         runTest {
-            val did1 = addDeck("Deck", setAsSelected = true).withNote(count = 2)
+            val did1 = addDeck("Deck", setAsSelected = true).withNotes(count = 2)
             val reviewReminderDeckSpecific = createAndSaveDummyDeckSpecificReminder(did1)
             val reviewReminderAppWide = createAndSaveDummyAppWideReminder()
 
@@ -243,7 +243,7 @@
     @Test
     fun `onReceive with rev cards not counted and only rev cards present should not fire notification`() =
         runTest {
-            val did1 = addDeck("Deck", setAsSelected = true).withNote(count = 2, queueType = QueueType.Rev)
+            val did1 = addDeck("Deck", setAsSelected = true).withNotes(count = 2, queueType = QueueType.Rev)
             val reviewReminder = createTestReminder(deckId = did1, countRev = false)
             ReviewRemindersDatabase.storeReminders(reviewReminder)
 
@@ -255,7 +255,7 @@
     @Test
     fun `onReceive with new cards not counted and only new cards present should not fire notification`() =
         runTest {
-            val did1 = addDeck("Deck").withNote(count = 2)
+            val did1 = addDeck("Deck").withNotes(count = 2)
             val reviewReminder = createTestReminder(deckId = did1, countNew = false)
             ReviewRemindersDatabase.storeReminders(reviewReminder)
 
@@ -267,7 +267,7 @@
     @Test
     fun `onReceive with lrn cards not counted and only lrn cards present should not fire notification`() =
         runTest {
-            val did1 = addDeck("Deck").withNote(count = 2, queueType = QueueType.Lrn)
+            val did1 = addDeck("Deck").withNotes(count = 2, queueType = QueueType.Lrn)
             val reviewReminder = createTestReminder(deckId = did1, countLrn = false)
             ReviewRemindersDatabase.storeReminders(reviewReminder)
 
@@ -319,8 +319,8 @@
     @Test
     fun `snooze actions of different notifications and different intervals should be different`() =
         runTest {
-            val did1 = addDeck("Deck1").withNote(count = 2)
-            val did2 = addDeck("Deck2").withNote(count = 2)
+            val did1 = addDeck("Deck1").withNotes(count = 2)
+            val did2 = addDeck("Deck2").withNotes(count = 2)
             val reviewReminderOne = createTestReminder(deckId = did1, thresholdInt = 1)
             val reviewReminderTwo = createTestReminder(deckId = did2, thresholdInt = 1)
             ReviewRemindersDatabase.storeReminders(reviewReminderOne, reviewReminderTwo)
Index: libanki/testutils/src/main/java/com/ichi2/anki/libanki/testutils/AnkiTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libanki/testutils/src/main/java/com/ichi2/anki/libanki/testutils/AnkiTest.kt b/libanki/testutils/src/main/java/com/ichi2/anki/libanki/testutils/AnkiTest.kt
--- a/libanki/testutils/src/main/java/com/ichi2/anki/libanki/testutils/AnkiTest.kt	(revision f4af44f18bcf1ad2a3348d3ec51a9e8d3b314072)
+++ b/libanki/testutils/src/main/java/com/ichi2/anki/libanki/testutils/AnkiTest.kt	(date 1768720430579)
@@ -226,10 +226,18 @@
     /**
      * Convenience method for chaining [addDeck] and [addNoteToDeck].
      *
-     * Usage: `val deckId = addDeck("My Deck").withNote(count = 5, queueType = QueueType.New)`
+     * Usage: `val deckId = addDeck("My Deck").withNote(queueType = QueueType.New)`
      */
-    fun DeckId.withNote(
-        count: Int = 1,
+    fun DeckId.withNote(queueType: QueueType = QueueType.New): DeckId
+        = withNotes(count = 1, queueType = queueType)
+
+    /**
+     * Convenience method for chaining [addDeck] and [addNoteToDeck].
+     *
+     * Usage: `val deckId = addDeck("My Deck").withNotes(count = 5, queueType = QueueType.New)`
+     */
+    fun DeckId.withNotes(
+        count: Int,
         queueType: QueueType = QueueType.New,
     ): DeckId =
         this.apply {
```

david-allison reviewed (approved):
Excellent, thanks so much!!

bhavESH-CHAWLA commented:
No problem ! Thanks for responding back @ericli3690

ericli3690 commented:
Working on a major refactor of the database! Then this guy will be rebased on top

ericli3690 commented:
Blocked behind:
- https://github.com/ankidroid/Anki-Android/pull/20325

ericli3690 commented:
Dependency merged; will revise soon, please give me a few days.
